pegasus_wrapper/__init__.py

```diff
@@ -1,2 +1,152 @@
+
+    Particular resource requests are implemented by cluster-specific sub-classes
+    of `ResourceRequest`.
+    """
+
+    num_cpus: int
+    num_gpus: int
+    memory: MemoryAmount
+    partition: str
+
+    @abstractmethod
+    def apply_to_job(self, job: Job, log_base_directory: Path, job_name: Optional[str] = "") -> None:
+        """
+        Applies the appropriate settings to *job*
+        to account for the requested resources.
+        """
+
+
+SLURM_RESOURCE_STRING = """--{qos_or_account} --partition {partition} --ntasks 1
+ --cpus-per-task {num_cpus} --gpus-per-task {num_gpus} --job-name {job_name} --mem {mem_str}
+ --output={stdout_log_path}"""
+
+
+@attrs(frozen=True, slots=True)
+class SlurmResourceRequest(ResourceRequest):
+    """
+    A `ResourceRequest` for a job running on a SLURM cluster.
+    """
+
+    memory: MemoryAmount = attrib(validator=instance_of(MemoryAmount), kw_only=True)
+    partition: str = attrib(validator=instance_of(str), kw_only=True)
+    num_cpus: int = attrib(validator=in_(Range.at_least(1)), default=1, kw_only=True)
+    num_gpus: int = attrib(validator=in_(Range.at_least(0)), default=0, kw_only=True)
+
+    def apply_to_job(
+        self, job: Job, log_base_directory: Path, job_name: str = "SlurmResourceJob"
+    ) -> None:
+        qos_or_account = (
+            f"qos {self.partition}"
+            if self.partition in ("scavenge", "ephemeral")
+            else f"account {self.partition}"
+        )
+        job_name += ".log"
+        slurm_resource_content = SLURM_RESOURCE_STRING.format(
+            qos_or_account=qos_or_account,
+            partition=self.partition,
+            num_cpus=self.num_cpus,
+            num_gpus=self.num_gpus,
+            job_name=job_name,
+            mem_str=to_slurm_memory_string(self.memory),
+            stdout_log_path=log_base_directory / job_name
+        )
+        job.add_profiles(
+            Profile(Namespace.PEGASUS, "glite.arguments", slurm_resource_content)
+        )
+
+
+@attrs(frozen=True, slots=True)
+class WorkflowBuilder:
+    """
+    A class which constructs a representation of a Pegasus workflow use `execute(workflow)` to build the workflow
+    """
+    name: str = attrib(validator=instance_of(str), kw_only=True)
+    created_by: str = attrib(validator=instance_of(str), kw_only=True)
+    log_base_directory: Path = attrib(validator=instance_of(Path), kw_only=True)
+    files_to_created_by_jobs: Dict[File, ImmutableSet[Job]] = attrib(kw_only=True, init=False, default=Factory(dict))
+
+    def schedule_job(self, job: Job, resource_request: ResourceRequest, *, job_name: Optional[str] = None) -> DependencyNode:
+        """
+        The provided Pegasus Job is expected to already have it's input files and output files linked. We may want to
+        wrap building a pegasus job somehow?
+
+        Ryan can you explain why you wanted to return a dependency node here? I'm not sure how that's helpful...
```

Comment:
I *think* we want `DependencyNode` to just wrap `Job` without necessarily exposing to the user that that is what we are doing.

The reason you want to return something here is so that the user can make `Artifact`s wrapping output objects, so that functions which are applied to those output objects can be aware of what job created them in order to set up the dependencies correctly.

Reply:
I tentatively think we want to provide a wrapper for `Job` which is similar but instead of working in terms of `File`s works in terms of `Artifacts` (where we for now throw an exception if any `Artifact` wraps something other than a `File`).  When the job-wrapper is passed to `schedule_job`, we can automatically unwrap the `Artifact`s to get the inputs to pass to the underlying job, and can also query those `Artifact`s to set the dependencies correctly (e.g. `diamond.depends(parent=preprocess, child=frl)` from the example).